File: leetcode/Validate_Binary_Search_Tree.java
import java.util.ArrayList;
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    boolean isOK(TreeNode r, int l, int h) {
        if (r == null)
            return true;
        if (r.val <= l || r.val >= h)
            return false;
        return (isOK(r.left, l, r.val) && isOK(r.right, r.val, h));
    }
    public boolean isValidBST(TreeNode root) {
        return isOK(root, -987654321, 987654321);
    }
}

File: leetcode/Validate_Binary_Search_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    bool check(TreeNode *node, int lower_bound, int upper_bound) {
        if (node == NULL)
            return true;
        // strictly less or greater than
        if (node->val <= lower_bound || node->val >= upper_bound)
            return false;
        if (!check(node->left,  lower_bound, node->val) ||
            !check(node->right, node->val,   upper_bound))
            return false;
        return true;
    }
public:
    bool isValidBST(TreeNode *root) {
        return check(root, -987654321, 987654321);
    }
};
